File: 107binaryTreeLevelOrderTraversalII/binaryTreeLevelOrderTraversalII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

const lodash = require('lodash')

Array.prototype.flatMap = function(callBack)
{
    return lodash.flatMap(this, callBack)
}

var levelOrderBottom = function(root) {
    const result = []
    let level = [root]
    while (root && level.length)
    {
        result.push(level.map(node => node.val))
        level = level.flatMap(node => [node.left, node.right].filter(_ => _))
    }
    return result.reverse()
};

File: 107binaryTreeLevelOrderTraversalII/binaryTreeLevelOrderTraversalII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        final var result = new ArrayList<List<Integer>>();
        var level = Arrays.asList(root);
        while (Objects.nonNull(root) && !level.isEmpty())
        {
            result.add(level.stream().map(node -> node.val).collect(Collectors.toList()));
            level = level.stream().flatMap(node -> Stream.of(node.left, node.right).filter(Objects::nonNull)).collect(Collectors.toList());
        }
        Collections.reverse(result);
        return result;
    } 
}
